File: Google/106.construct-binary-tree-from-inorder-and-postorder-traversal.java
import java.util.*;

/**
 * Algorithm:
 * Postorder gives the root last; inorder splits the subtree around that root.
 * Recursively build left and right ranges using a value-to-index map.
 *
 * Java data structures:
 * HashMap<Integer, Integer> avoids scanning inorder for every root.
 *
 * Complexity:
 * Time O(n), space O(n) for the map plus O(h) recursion stack.
 */
class Solution {
    private int[] postorder;
    private Map<Integer, Integer> index;

    public TreeNode buildTree(int[] inorder, int[] postorder) {
        this.postorder = postorder;
        index = new HashMap<>();
        for (int i = 0; i < inorder.length; i++) {
            index.put(inorder[i], i);
        }
        return build(0, inorder.length - 1, 0, postorder.length - 1);
    }

    private TreeNode build(int inL, int inR, int postL, int postR) {
        if (inL > inR) {
            return null;
        }
        int rootVal = postorder[postR];
        int mid = index.get(rootVal);
        int leftSize = mid - inL;
        TreeNode root = new TreeNode(rootVal);
        root.left = build(inL, mid - 1, postL, postL + leftSize - 1);
        root.right = build(mid + 1, inR, postL + leftSize, postR - 1);
        return root;
    }
}

File: companies/Microsoft/106.construct-binary-tree-from-inorder-and-postorder-traversal.go
package leetcode

// BuildTree106 reconstructs a tree from inorder and postorder traversals.
// Postorder gives each subtree root last; the inorder index splits left and
// right subtree ranges.
//
// Time: O(n). Space: O(n).
func BuildTree106(inorder []int, postorder []int) *TreeNode {
	idx := map[int]int{}
	for i, v := range inorder {
		idx[v] = i
	}
	var build func(int, int, int, int) *TreeNode
	build = func(inL, inR, postL, postR int) *TreeNode {
		if inL > inR {
			return nil
		}
		rootVal := postorder[postR]
		mid := idx[rootVal]
		leftSize := mid - inL
		root := &TreeNode{Val: rootVal}
		root.Left = build(inL, mid-1, postL, postL+leftSize-1)
		root.Right = build(mid+1, inR, postL+leftSize, postR-1)
		return root
	}
	return build(0, len(inorder)-1, 0, len(postorder)-1)
}
